### PR TITLE
[CAS] Teach swift compiler to compute cache key and store outputs into CAS

### PR DESCRIPTION
**Pull request by cachemeifyoucan**
Teach swift compiler how to compute cache key and store outputs into CAS. It contains:

* Cache key computation logic
* CAS output backend that mirrors the output into CAS
* Add cache lookup entry for outputs in ActionCache
* A swift-driver tool that can be used to inspect caching functions.

### CODE REVIEW

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test macOS platform

File: test/CAS/cache_key_compute.swift
Comment (benlangmuir) [inline test/CAS/cache_key_compute.swift:5]: should we check the error message?

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (benlangmuir) [inline lib/Frontend/CompileJobCacheKey.cpp:57]: This should report the Error. Same with the other cantFail below.

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (benlangmuir) [inline lib/Frontend/CompileJobCacheKey.cpp:27]: How are input file contents accounted for here?

File: lib/DriverTool/swift_cache_tool_main.cpp
Comment (benlangmuir) [inline lib/DriverTool/swift_cache_tool_main.cpp:92]: Help should probably return 0.

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (cachemeifyoucan) [reply lib/Frontend/CompileJobCacheKey.cpp:27]: It is expected to come from `-cas-fs` argument. Currently, swift-frontend will produce the cache key even when `-cas-fs` is not used. This is risky since file content are in the cache key and the cache entry can't really be used but on the other hand, it makes writing unit test easier in swift repo since compute `cas-fs` for a swift invocation is not easy.

Comment (benlangmuir) [reply lib/Frontend/CompileJobCacheKey.cpp:27]: Makes sense. How about adding a warning if there is no cas-fs?  Would be good to explain this in a comment here too.  I'm fine with tests cheating here, just don't want to accidentally use this in production 😸

Comment (cachemeifyoucan) [reply lib/Frontend/CompileJobCacheKey.cpp:27]: True. Let me actually just add a hidden flag like `-allow-unstable-cache-key-for-test` which we can use in test, otherwise, it will just error out if caching is used without `-cas-fs`

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Review by benlangmuir — APPROVED** [review]
LGTM, but please wait for at least one more reviewer to approve since I'm not that familiar with the swift frontend.

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

File: include/swift/Frontend/CachingUtils.h
Comment (artemcm) [inline include/swift/Frontend/CachingUtils.h:45]: What about many-to-one cases, like WMO?

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (artemcm) [inline lib/Frontend/CompileJobCacheKey.cpp:33]: Much like we have `DoesNotAffectIncrementalBuild`, we can add a `DoesNotContributeToCASKey` flag to argument definitions, in the future.

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (artemcm) [inline lib/Frontend/CompileJobCacheKey.cpp:39]: Is this not mutating the same underlying storage as was passed to `swift::performFrontend`?

File: lib/Frontend/Frontend.cpp
Comment (artemcm) [inline lib/Frontend/Frontend.cpp:473]: Can this be made work with `ParsedArgs`? 
Previously, downstream of  `parseArgs`, the rest of the compiler does not consult the raw argument values, for several good reasons (e.g. de-duplication). 
I think there's value to keeping that convention.

File: lib/Frontend/CompileJobCacheKey.cpp
Comment (cachemeifyoucan) [reply lib/Frontend/CompileJobCacheKey.cpp:33]: Sounds like a good idea but we actually would love to have argument canonicalization so we can serialize/deserialize CompilerInvocation for caching purpose.

File: lib/Frontend/Frontend.cpp
Comment (cachemeifyoucan) [reply lib/Frontend/Frontend.cpp:473]: Maybe ParsedArgs is better, but it has to be render backed into raw argument value for cache key currently. ParsedArg doesn't de-duplicate or handle ordering difference. The ideal solution is doing a canonicalization for CompilerInvocation

Comment (artemcm) [reply lib/Frontend/Frontend.cpp:473]: Okay then, thanks, let's leave it as-is for now, but these details are adding up to make canonicalization increasingly important. 👍🏼

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Comment by cachemeifyoucan** [comment]
@swift-ci please smoke test

**Review by artemcm — APPROVED** [review]

**Pull request merged**
